### PR TITLE
[fix] recover auto-dispatch for rmsnorm and rope

### PR DESCRIPTION
**Pull request by Alcanderian**
## Checklist

- [ ] Format your code according to the [Code Formatting with Pre-Commit](https://docs.sglang.ai/references/contribution_guide.html#code-formatting-with-pre-commit).
- [ ] Add unit tests as outlined in the [Running Unit Tests](https://docs.sglang.ai/references/contribution_guide.html#running-unit-tests-adding-to-ci).
- [ ] Update documentation / docstrings / example tutorials as needed, according to [Writing Documentation](https://docs.sglang.ai/references/contribution_guide.html#writing-documentation-running-docs-ci).
- [ ] Provide throughput / latency benchmark results and accuracy evaluation results as needed, according to [Benchmark and Profiling](https://docs.sglang.ai/references/benchmark_and_profiling.html) and [Accuracy Results](https://docs.sglang.ai/references/accuracy_evaluation.html).
- [ ] For reviewers: If you haven't made any contributions to this PR and are only assisting with merging the main branch, please remove yourself as a co-author when merging the PR.
- [ ] Please feel free to join our Slack channel at https://slack.sglang.ai to discuss your PR.

### CODE REVIEW

File: python/sglang/srt/layers/layernorm.py [hunk 1]
```diff
@@ -52,12 +52,8 @@ def __init__(
     def forward(self, *args, **kwargs):
         if torch.compiler.is_compiling():
             return self.forward_native(*args, **kwargs)
-        if _is_cuda:
-            return self.forward_cuda(*args, **kwargs)
-        elif _is_hip:
-            return self.forward_hip(*args, **kwargs)
         else:
-            return self.forward_native(*args, **kwargs)
+            return self._forward_method(*args, **kwargs)
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/layernorm.py]: nit: wondering whether it is possible we just completely remove all the forward methods?

btw, this refactor may be related: https://github.com/sgl-project/sglang/pull/5382

Comment (Alcanderian) [reply python/sglang/srt/layers/layernorm.py]: > nit: wondering whether it is possible we just completely remove all the forward methods?
> 
> btw, this refactor may be related: #5382

Looks good! We can record the origin _forwrd_method in enter_torch_compile and recover it in leave_torch_compile, rather than recovering to forward_cuda

Comment (fzyzcjy) [reply python/sglang/srt/layers/layernorm.py]: Pretty reasonable

Comment (fzyzcjy) [reply python/sglang/srt/layers/layernorm.py]: what about firstly merging 5382, and then do a modifcation

I personally like to have a refactor PR + a change PR, o/w a PR with both refactor and change will make it pretty hard to trace

Comment (ispobock) [inline python/sglang/srt/layers/layernorm.py]: Should be careful with this change on AMD devices. CI test seems failed: https://github.com/sgl-project/sglang/actions/runs/15331204245/job/43138169215?pr=6745
Previously I fixed it in https://github.com/sgl-project/sglang/pull/5723. It seems that `_forward_method` will call the super class's `forward_hip` by default even though we override it.

Comment (Alcanderian) [reply python/sglang/srt/layers/layernorm.py]: Got it, I will find a better way to resolve this issue!

Comment (Alcanderian) [reply python/sglang/srt/layers/layernorm.py]: We need to redispatch _forward_method in RMSNorm because the super class cannot access sub class method. This issue will aslo appear at https://github.com/sgl-project/sglang/pull/5382. @fzyzcjy

Comment (Alcanderian) [reply python/sglang/srt/layers/layernorm.py]: We need to redispatch _forward_method at RMSNorm::__init__. Because python do not support override util 3.12. This issue will also appear in https://github.com/sgl-project/sglang/pull/5382 @fzyzcjy

Comment (fzyzcjy) [reply python/sglang/srt/layers/layernorm.py]: I see, looks reasonable (I will keep 5382 unchanged to avoid changing semantics during a refactor)

**Comment by Alcanderian** [comment]
@ispobock @fzyzcjy It works fine now! I would really appreciate it if you could take another look!

File: python/sglang/srt/custom_op.py [hunk 2]
```diff
@@ -10,8 +10,18 @@ class CustomOp(nn.Module):
...
         self._forward_method = self.dispatch_forward()
+        self.is_torch_compile = False
 
     def enter_torch_compile(self, num_tokens: int):
+        # Skip if Op is already entered compile mode.
+        # NOTE(alcanderian): Some Ops(for example RotaryEmbedding) will be reused
+        # among layers and `enter_torch_compile` will be called many times.
+        # We should prevent `self._origin_forward_method` from being overridden when
+        # it is not the first time `enter_torch_compile` called.
+        if self.is_torch_compile == True:
```
Comment (fzyzcjy) [inline python/sglang/srt/custom_op.py]: nit

```suggestion
        if self.is_torch_compile:
```

File: python/sglang/srt/custom_op.py [hunk 3]
```diff
@@ -27,7 +37,12 @@ def enter_torch_compile(self, num_tokens: int):
         self.is_torch_compile = True
 
     def leave_torch_compile(self):
-        self._forward_method = self.forward_cuda
+        # Skip if Op is already exited compile mode.
+        if self.is_torch_compile == False:
```
Comment (fzyzcjy) [inline python/sglang/srt/custom_op.py]: nit
```suggestion
        if not self.is_torch_compile:
```

File: python/sglang/srt/custom_op.py [hunk 4]
```diff
@@ -10,8 +10,18 @@ class CustomOp(nn.Module):
...
     def enter_torch_compile(self, num_tokens: int):
+        # Skip if Op is already entered compile mode.
+        # NOTE(alcanderian): Some Ops(for example RotaryEmbedding) will be reused
+        # among layers and `enter_torch_compile` will be called many times.
+        # We should prevent `self._origin_forward_method` from being overridden when
+        # it is not the first time `enter_torch_compile` called.
+        if self.is_torch_compile == True:
+            return
+
+        self._origin_forward_method = self._forward_method
```
Comment (fzyzcjy) [inline python/sglang/srt/custom_op.py]: nit: "origin" -> "original"

File: python/sglang/srt/custom_op.py [hunk 5]
```diff
@@ -10,8 +10,18 @@ class CustomOp(nn.Module):
     def __init__(self):
         super().__init__()
         self._forward_method = self.dispatch_forward()
+        self.is_torch_compile = False
```
Comment (fzyzcjy) [inline python/sglang/srt/custom_op.py:16]: nit: what about adding a `self._original_forward_method = None`

File: python/sglang/srt/layers/layernorm.py [hunk 6]
```diff
@@ -49,15 +49,9 @@ def __init__(
...
-            return self.forward_native(*args, **kwargs)
-        if _is_cuda:
-            return self.forward_cuda(*args, **kwargs)
-        elif _is_hip:
-            return self.forward_hip(*args, **kwargs)
-        else:
-            return self.forward_native(*args, **kwargs)
+        # Re-dispatch
+        if _is_hip:
+            self._forward_method = self.forward_hip
```
Comment (fzyzcjy) [inline python/sglang/srt/layers/layernorm.py]: wondering whether it is possible we do not treat RMSNorm etc to be special, i.e. make all ops treated in the same way?

**Review by fzyzcjy — COMMENTED** [review]
Only some nits (it is ok to ignore all of them) and a curious question

File: python/sglang/srt/layers/layernorm.py (hunk 6 above)
Comment (Alcanderian) [reply python/sglang/srt/layers/layernorm.py]: > wondering whether it is possible we do not treat RMSNorm etc to be special, but instead make it treated the same way as other ops?

Fundamentally, this is actually an issue caused by AMD. To fix it properly, we should probably modify the `forward_hip` of `CustomOp` and implement `forward_hip` for all other operators. Currently, this seems like quite significant modifications. Frankly speaking, I also strongly oppose specializing logic like [this](https://github.com/sgl-project/sglang/pull/6838/files#diff-190683f5c7ffde48eb0ce015d76532f833fe1e122779eae6da8181dbfab3895bR79).

Comment (Alcanderian) [reply python/sglang/srt/layers/layernorm.py]: > wondering whether it is possible we do not treat RMSNorm etc to be special, i.e. make all ops treated in the same way?

But seems fine to remove these now since torch.compile logic has been fixed.

**Review by fzyzcjy — APPROVED** [review]
LGTM (as long as all tests pass)

File: python/sglang/srt/layers/layernorm.py [hunk 7]
```diff
@@ -49,16 +49,6 @@ def __init__(
         self.weight = nn.Parameter(torch.ones(hidden_size))
         self.variance_epsilon = eps
 
-    def forward(self, *args, **kwargs):
```
Comment (HaiShaw) [inline python/sglang/srt/layers/layernorm.py:52]: we would still need this custom dispatch in this context - over-general abstraction as _is_hip to forward_hip isn't enough - we will need to add forward_aiter specifically in this block.

**Review by HaiShaw — COMMENTED** [review]
@fzyzcjy @Alcanderian @zhyncs 
some question to layernorm changes
